File: backend/app/schemas/capability.py
"""Schemas for the enterprise welding capability library."""
from typing import Any, Literal

from pydantic import BaseModel, Field


class CapabilityFilters(BaseModel):
    factory_id: int | None = Field(None, gt=0)
    process: str | None = Field(None, max_length=100)
    material_group: str | None = Field(None, max_length=100)
    position: str | None = Field(None, max_length=100)
    search: str | None = Field(None, max_length=200)


class CapabilityOverviewResponse(BaseModel):
    generated_at: str
    workspace: dict[str, Any]
    filters: dict[str, Any]
    summary: dict[str, int | float]
    health: dict[str, Any]
    dimensions: dict[str, list[Any]]
    wps_records: list[dict[str, Any]]
    pqr_records: list[dict[str, Any]]
    welders: list[dict[str, Any]]
    process_capabilities: list[dict[str, Any]]
    materials: list[dict[str, Any]]
    equipment: list[dict[str, Any]]
    issues: list[dict[str, Any]]


class CapabilityCheckRequest(BaseModel):
    standard_system: Literal["china", "asme", "ped"] = "china"
    factory_id: int | None = Field(None, gt=0)
    welding_process: str = Field(min_length=1, max_length=100)
    material_group: str = Field(min_length=1, max_length=100)
    thickness_mm: float = Field(gt=0, le=10000)
    diameter_mm: float | None = Field(None, gt=0, le=100000)
    welding_position: str = Field(min_length=1, max_length=100)
    pwht_required: bool = False
    impact_required: bool = False
    impact_temperature_c: float | None = Field(None, ge=-273.15, le=2000)


class CapabilityCheckResponse(BaseModel):
    decision: str
    process_capable: bool
    personnel_capable: bool
    resource_ready: bool
    requirement: dict[str, Any]
    matched_capabilities: list[dict[str, Any]]
    matched_welders: list[dict[str, Any]]
    matched_materials: list[dict[str, Any]]
    matched_equipment: list[dict[str, Any]]
    gaps: list[dict[str, Any]]
    explanation: list[str]

File: frontend/src/services/capability.ts
import apiService from './api'

export interface CapabilityFilters {
  factory_id?: number
  process?: string
  material_group?: string
  position?: string
  search?: string
}

export interface CapabilityOverview {
  generated_at: string
  workspace: { type: string; company_id?: number; factory_id?: number }
  filters: CapabilityFilters
  summary: Record<string, number>
  health: {
    score: number
    status: 'healthy' | 'attention' | 'risk'
    blocking_issue_count: number
    warning_count: number
    unsupported_wps_count: number
    expiring_welder_count: number
    valid_relation_count: number
  }
  dimensions: {
    processes: string[]
    material_groups: string[]
    positions: string[]
    thickness_ranges: Array<Record<string, unknown>>
    diameter_ranges: Array<Record<string, unknown>>
    pwht_conditions: boolean[]
    impact_conditions: boolean[]
  }
  wps_records: Array<Record<string, any>>
  pqr_records: Array<Record<string, any>>
  welders: Array<Record<string, any>>
  process_capabilities: Array<Record<string, any>>
  materials: Array<Record<string, any>>
  equipment: Array<Record<string, any>>
  issues: Array<Record<string, any>>
}

export interface CapabilityCheckRequest {
  factory_id?: number
  welding_process: string
  material_group: string
  thickness_mm: number
  diameter_mm?: number
  welding_position: string
  pwht_required: boolean
  impact_required: boolean
  impact_temperature_c?: number
}

export interface CapabilityCheckResult {
  decision: 'capable' | 'needs_resources' | 'not_capable'
  process_capable: boolean
  personnel_capable: boolean
  resource_ready: boolean
  requirement: CapabilityCheckRequest
  matched_capabilities: Array<Record<string, any>>
  matched_welders: Array<Record<string, any>>
  matched_materials: Array<Record<string, any>>
  matched_equipment: Array<Record<string, any>>
  gaps: Array<Record<string, any>>
  explanation: string[]
}

class CapabilityService {
  async getOverview(filters: CapabilityFilters = {}): Promise<CapabilityOverview> {
    const response = await apiService.get<CapabilityOverview>(
      '/qualification/capabilities/overview',
      { params: filters }
    )
    return response.data
  }

  async check(request: CapabilityCheckRequest): Promise<CapabilityCheckResult> {
    const response = await apiService.post<CapabilityCheckResult>(
      '/qualification/capabilities/check',
      request
    )
    return response.data
  }
}

export const capabilityService = new CapabilityService()
